modules/nexverseBidAdapter.js

```diff
@@ -0,0 +1,447 @@
+/* eslint-disable camelcase */
+
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+import { BANNER, VIDEO, NATIVE } from '../src/mediaTypes.js';
+import { isArray, logError, logWarn, logInfo } from '../src/utils.js';
+
+const BIDDER_CODE = 'nexverse';
+const ENDPOINT_URL = 'https://rtb.nexverse.ai/';
+const SUPPORTED_MEDIA_TYPES = [BANNER, VIDEO, NATIVE];
+
+const LOG_WARN_PREFIX = '[Nexverse warn]: ';
+const LOG_ERROR_PREFIX = '[Nexverse error]: ';
+const LOG_INFO_PREFIX = '[Nexverse info]: ';
+
+export const spec = {
+  code: BIDDER_CODE,
+  supportedMediaTypes: SUPPORTED_MEDIA_TYPES,
+
+  /**
+   * Validates the bid request to ensure all required parameters are present.
+   * @param {Object} bid - The bid request object.
+   * @returns {boolean} True if the bid request is valid, false otherwise.
+   */
+  isBidRequestValid(bid) {
+    const isValid = !!(bid.params && bid.params.uid && bid.params.pub_id && bid.params.pub_epid);
+
+    if (!isValid) {
+      logError(`${LOG_ERROR_PREFIX} Missing required bid parameters.`);
+    }
+
+    return isValid;
+  },
+
+  /**
+   * Builds the OpenRTB server request from the list of valid bid requests.
+   *
+   * @param {Array} validBidRequests - Array of valid bid requests.
+   * @param {Object} bidderRequest - The bidder request object containing additional data.
+   * @returns {Array} Array of server requests to be sent to the endpoint.
+   */
+  buildRequests(validBidRequests, bidderRequest) {
+    const requests = validBidRequests.map((bid) => {
+      const { uid, pub_id, pub_epid } = bid.params;
+
+      if (!uid || !pub_id || !pub_epid) {
+        logError(`${LOG_ERROR_PREFIX} Missing required endpoint URL parameters.`);
+        return null; // Stop processing this bid
+      }
+
+      // Build the endpoint URL with query parameters
+      const endpointUrl = `${ENDPOINT_URL}?uid=${encodeURIComponent(uid)}&pub_id=${encodeURIComponent(
+        pub_id
+      )}&pub_epid=${encodeURIComponent(pub_epid)}`;
+
+      // Build the OpenRTB payload
+      const payload = buildOpenRtbRequest(bid, bidderRequest);
+
+      if (!payload) {
+        logError(`${LOG_ERROR_PREFIX} Payload could not be built.`);
+        return null; // Skip this bid
+      }
+
+      // Return the server request
+      return {
+        method: 'POST',
+        url: endpointUrl,
+        data: JSON.stringify(payload),
+        bidRequest: bid,
+        options: {
+          contentType: 'application/json',
```

Comment:
This slows things down and forces a cors options call

Reply:
Fixed in new PR https://github.com/prebid/Prebid.js/pull/12472

Reply:
Done.